### DISCUSSION TITLE
YAML environment variables

### DISCUSSION THREAD

**Discussion by rennokki** [laravel/framework#42283] [Ideas]
### The problem

This is strongly linked to a type of problem I tried to implement using .env variables, but I ended up writing my own class to do it in order to achieve it. Here it is how it goes:

Let's assume you want to create a config file that holds connection details for multiple clusters (it can be anything; conceptually) and looks like this after compiling with `php artisan config:cache` (where details like `host` should be different between environments):

```php
// config/clusters.php

return [

    'clusters' => [
        'us-east-1' => [
            'host' => 'prod-cluster-host.us-east-1.amaz...',
            // ...
        ],
        'eu-central-1' => [
            'host' => 'prod-cluster-host.eu-central-1.amaz...',
            // ...
        ],
        // ...
    ],

];
```

This can be ugly to achieve with `.env` variables, like this (using the multi-line variables):

```php
// config/clusters.php

return [

    'clusters' => json_decode(env('CLUSTERS', '{}'), true),

];
```

```bash
CLUSTERS='{
  "us-east-1": {
    "host": "prod-cluster-host.us-east-1.amaz..."
  },
  "eu-central-1": {
    "host": "prod-cluster-host.eu-central-1.amaz..."
  }
}'
```

### .env.yaml file

For complex, object-like variables, a YAML configuration file would be better:

```php
// config/clusters.php

return [

    'clusters' => yaml_env('CLUSTERS', []),

];
```

```yaml
# .env.yaml or smth like that

CLUSTERS:
  us-east-1:
    host: prod-cluster-host.us-east-1.amaz...
  eu-central-1:
    host: prod-cluster-host.eu-central-1.amaz...
```

## Custom implementation problems

With custom implementations, you cannot use the config files' caching, as the only issue is that the `php artisan config:cache` would not cache anything, so in case you would want to implement this feature in every project you need, you will need to find a way to cache it and invalidate it, increasing the complexity.

## Existing configuration

Since existing configuration is "there are defaults, change only what you need", using YAML would make no use most of the times. When configuring a driver for the database, for example, you can declare the variables in YAML the same way you would do with .env files, having ambiguity in case of debugging complex configurations, sine the are multiple files where variables can be defined:

```bash
DB_HOST=10.0.0.1
```

```yaml
DB_HOST: 10.0.0.1
```

YAML format would make sense for custom-made or define complex configurations that would otherwise not look better in JSON-format, especially for the fact that .env syntax highlighting would not be working the same way as a separate YAML file.

**Comment by rw4lll** [comment]
I would really appreciate to have this feature globally. Yaml is very common format and it would be great to have a choice between usage env or yaml in the project.

**Reply by fatrbaby** [reply]
> I would really appreciate to have this feature globally. Yaml is very common format and it would be great to have a choice between usage env or yaml in the project.

Please see:  https://12factor.net/config 

Can choose between `PHP` and `yaml`, but you must not choose between `env` and `yaml`.

**Comment by rennokki** [comment]
I'm trying to build a package for this (https://github.com/renoki-co/laravel-yaml-env). I'm still deciding if I should do:

## For ENV

```php
// config/aws.php
return [
    'clusters' => env('CLUSTERS')
];
```

```yaml
CLUSTERS:
  - region: us-east-1
  - region: us-east-2
```

## For config

```php
// config/aws.php
return [
    //
];
```

```yaml
aws: # the first key is the config name
  clusters:
    - region: us-east-1
    - region: us-east-2
```
